compiler/grainformat/reformat.re

```diff
@@ -0,0 +1,2753 @@
+      )
+    | PPatArray(patterns) => (
+        Doc.group(
+          Doc.concat([
+            Doc.lbracket,
+            Doc.text(">"),
+            Doc.space,
+            Doc.join(
+              Doc.concat([Doc.comma, Doc.space]),
+              List.map(
+                p => print_pattern(~pat=p, ~parent_loc, ~original_source),
+                patterns,
+              ),
+            ),
+            Doc.rbracket,
+          ]),
+        ),
+        false,
+      )
+    | PPatRecord(patternlocs, closedflag) => (
+        print_record_pattern(
+          ~patternlocs,
+          ~closedflag,
+          ~parent_loc,
+          ~original_source,
+        ),
+        false,
+      )
+    | PPatConstraint(pattern, parsed_type) => (
+        Doc.concat([
+          print_pattern(
+            ~pat=pattern,
+            ~parent_loc=pat.ppat_loc,
+            ~original_source,
+          ),
+          Doc.concat([Doc.text(":"), Doc.space]),
+          print_type(parsed_type),
+        ]),
+        false,
+      )
+    | PPatConstruct(location, patterns) =>
+      let func =
+        switch (location.txt) {
+        | IdentName(name) => name
+        | _ => ""
+        };
+      if (func == "[...]") {
+        (
+          resugar_list_patterns(~patterns, ~parent_loc, ~original_source),
+          false,
+        );
+      } else {
+        (
+          Doc.concat([
+            print_ident(location.txt),
+            if (List.length(patterns) > 0) {
+              add_parens(
+                Doc.join(
+                  Doc.comma,
+                  List.map(
+                    pat => print_pattern(~pat, ~parent_loc, ~original_source),
+                    patterns,
+                  ),
+                ),
+              );
+            } else {
+              Doc.nil;
+            },
+          ]),
+          false,
+        );
+      };
+
+    | PPatOr(pattern1, pattern2) => (
+        Doc.text("/* Formatter PPatOr error */"),
+        false,
+      )
+    | PPatAlias(pattern, loc) => (
+        Doc.text("/* Formatter PPatAlias error */"),
+        false,
+      )
+    };
+
+  let (pattern, parens) = printed_pattern;
+
+  let with_leading =
+    if (leading_comment_docs == Doc.nil) {
+      [pattern];
+    } else {
+      [leading_comment_docs, Doc.space, pattern];
+    };
+  let with_trailing =
+    if (trailing_comment_docs == Doc.nil) {
+      with_leading;
+    } else {
+      List.append(with_leading, [trailing_comment_docs]);
+    };
+
+  let clean_pattern =
+    if (List.length(with_trailing) == 1) {
+      List.hd(with_trailing);
+    } else {
+      Doc.concat(with_trailing);
+    };
+
+  if (parens) {
+    Doc.concat([Doc.lparen, clean_pattern, Doc.rparen]);
+  } else {
+    clean_pattern;
+  };
+}
+and print_constant = (c: Parsetree.constant) => {
+  let print_c =
+    switch (c) {
+    | PConstNumber(PConstNumberInt(i)) => Printf.sprintf("%s", i)
+    | PConstNumber(PConstNumberFloat(f)) => Printf.sprintf("%s", f)
```

Comment:
Do we want to do any reformatting of numbers? Remove/add trailing zeros for a float, etc?

Reply:
FWIW I'd like to preserve the form a number is written in, i.e. `0x77` should stay in hex. (Looks like this is the case, but I figure I'd call it out if we were to make other changes.)

Reply:
Confirmed hex values print through and I've added tests for all the other value types too